### ISSUE TITLE
Remove weak dependencies

### ISSUE THREAD

**Issue by manuelbb-upb** [LuxDL/Lux.jl#294]
Big fan of this project, thanks a lot!

I noticed that installing/precompiling takes a while and that there are quite some heavy dependencies.
By looking at LuxCUDA.jl, it seems that CUDA is going to be optional?
 
However, there is more:
From how I interpret the [`__init__`](https://github.com/LuxDL/Lux.jl/blob/b292c45ddaf21f134cb1676e5f8526670e4cf7fc/src/Lux.jl#L59) function, it appears that Flux, Zygote etc. are meant to be optional dependencies. 
But they appear in the `[deps]` section of the Project.toml and are automatically installed in Julia 1.8.
I did not test Julia 1.9 yet, but as they are also listed in `[weakdeps]`, it might not be the case for Julia 1.9+

Is it mandatory to have Flux, ... as hard dependencies? Or could they be removed from the Project.toml dependency section?

[This issue](https://github.com/LuxDL/Lux.jl/issues/45) also mentions Flux as a dependency, so might be related.

**Comment by avik-pal** [comment]
>  By looking at LuxCUDA.jl, it seems that CUDA is going to be optional?

Yes, CUDA is going to be optional starting v0.5 (no planned timeline for it yet).

> it appears that Flux, Zygote etc. are meant to be optional dependencies.

Correct. However, for the Project.toml to be backwards compatible, we need to list all dependencies in `[deps]` and `[weakdeps]` (See [DiffEqBase.jl](https://github.com/SciML/DiffEqBase.jl/blob/e8c0a33c5a6be7d7bbbca190e7e0a14d5a5f7f1d/Project.toml) as an example) (I could be wrong here since weak deps is quite new and I am not very familiar with it). Which means that you will pay the price for installing all the packages, however, those should never be loaded. If you `instantiate` and do `using Lux` only the required packages will be precompiled.

**Comment by manuelbb-upb** [comment]
Thanks for the clarification :+1: You are right, I found out where I triggered precompilation of Zygote myself.

In DiffEqBase, most of the weakdeps do not appear in the `[deps]` section (except for Distributions, which might be accidental).
What I don't quite understand yet: Why do the weak dependencies have to be installed for backwards compatibility? From how they are setup with Requires.jl, everything should just work fine. (And they can have compat entries without being listed in `[deps]`, [it seems](https://github.com/JuliaPackaging/Requires.jl/blob/542a641b633fa38a5950c2b1dbb90861967c7015/README.md?plain=1#L61))

If removing the weak dependencies from `[deps]` is too much of a braking change, then we/you could maybe evaluate this as a task for 0.5?

**Comment by avik-pal** [comment]
Ah I see. I think we can get away with just putting them under `[extras]`. We need it to be somewhere in Project.toml since we need `[compat]` entries.

**Comment by avik-pal** [comment]
If the tests pass for the linked PR this should be fixed
